_posts/2020-01-30-database-migration-spring-boot-flyway.md

```diff
@@ -0,0 +1,134 @@
+* How a new database instance can be set up to specific or latest state?
+
+Answering these questions required one of us to check for a change which was part of the SQL script, e.g. adding a column, a stored procedure etc. If we multiply all these checks to the number of environments plus the work required to align the state, then we get a decent amount of time lost.
+
+Database migrations allow you to:
+
+* Create a database from scratch.
+* Have a single source of truth for database version.
+* Have reproducible state of the database in local and remote environments.
+* Automate database changes deployment, which helps to minimize human errors.
+
+### Why Flyway
+
+Flyway facilitates the above while providing:
+
+1. Well structured and easy to read documentation.
+2. An option to integrate with an existing database.
+3. Support for almost all known databases.
+4. Wide variety of running and configuration options.
+
+## Setting up Flyway
+
+We use H2 database in an `in-memory` mode for this post. Spring Boot auto-configures [Flyway](https://docs.spring.io/spring-boot/docs/current/reference/html/howto.html#howto-execute-flyway-database-migrations-on-startup) and [H2](https://docs.spring.io/spring-boot/docs/current/reference/html/spring-boot-features.html#boot-features-embedded-database-support) as long as we add the following dependencies to our build file (Gradle notation):
+
+```
+implementation 'org.flywaydb:flyway-core'
+runtimeOnly 'com.h2database:h2:1.4.199'
+```
+
+This is enough for Spring Boot to initializes a specific implementation of `DataSource` called `EmbeddedDatabase`. It is  then used by Flyway to execute the given migrations. All configuration options can be found at [Spring Boot application properties reference](https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-application-properties.html)
+
+In case we prefer to use one of the build tool plugins, an additional definition is required in our build file (Gradle notation):
+
+```gradle
+plugins {
+
+  // Other plugins...
+ 
+  id "org.flywaydb.flyway" version "6.2.0"
+}
+
+flyway {
+  url = 'jdbc:h2:mem:'
+}
+```
+
+## Writing Our First Database Migration
+
+Flyway has own [naming convention](https://flywaydb.org/documentation/migrations#naming) which can be adjusted to our needs using the following [configurations](https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-application-properties.html#data-migration-properties):
+
+```
+spring.flyway.sql-migration-prefix
+spring.flyway.sql-migration-separator
+spring.flyway.sql-migration-suffixes
+```
+
+Here is the `V1__init.sql` file used in our [code examples repository](https://github.com/thombergs/code-examples/tree/master/spring-boot) **TODO: Put actual link here when the code is approved**:
+
+```sql
+CREATE TABLE auth_user(
+    id INT AUTO_INCREMENT PRIMARY KEY,
+    username VARCHAR(255) NOT NULL unique
+);
+```
+
+## Running Flyway
```

Comment:
I do agree it should have separate sections, but to limit it to:
1. Automatic on startup
2. Gradle plugin
as mostly used across developers and closer to Spring Boot
I don't mind describing each of the options listed, but wouldn't this repeat the information provided in Flyway documentation?

Reply:
We can leave the Docker option out, that's fine.

Yes, this would "repeat" information from the Flyway documentation, but in your own words, and in another context. 

We don't want to just repeat everything in the documentation, but we also don't want just to link to the documentation in every other sentence. The goal is to provide a single source of information that is good enough to get the reader started. If the reader still needs more information, they can follow the link to the reference documentation :).